lib/http_get.js

```diff
@@ -0,0 +1,312 @@
+ * @param {Object} errorAgent - Agent to persist errors
+ * @param {Object} errorObj - Description of errors to be persisted
+ * @param {HyperdriveClient~deleteCallback} callback - Callback
+ * @param {null|Error} currentError - current final status of operation
+ * @param {boolean} callClientCb - Should we invoke callback?
+ * @return {Any} whatever callback returns
+ */
+function _handleErrors(opContext, errorAgent, errorObj,
+                       callback, currentError, callClientCb) {
+    const payload = {
+        topic: httpUtils.topics.repair,
+        key: opContext.fragments.objectKey,
+        messages: [JSON.stringify(errorObj)],
+    };
+
+    errorAgent.send([payload], err => {
+        if (err) {
+            opContext.log.error(
+                'Failure to persist fragments to repair', err);
+            // Return persistence error
+            // TODO: not sure this a proper HTTP error....
+            // TODO: must be a better way than disabling it
+            /* eslint-disable no-param-reassign */
+            err.infos = { status: 500, method: 'GET' };
+            opContext.failedToPersist = true;
+            /* eslint-enable no-param-reassign */
+            return callClientCb ? callback(err) : null;
+        }
+
+        /* Error persistence OK, return already encountered
+         * error (might be null) */
+        return callClientCb ? callback(currentError) : null;
+    });
+}
+
+function _handleCorruption(reqContext, corruptError, errorAgent) {
+    const toRepair = {
+        rawKey: reqContext.opContext.rawKey,
+        fragments: [[reqContext.chunkId, reqContext.fragmentId]],
+    };
+    _handleErrors(reqContext.opContext, errorAgent, toRepair,
+                  null, corruptError, false);
+}
+
+function getCorruptionCheckedStream(httpReply, reqContext, errorAgent) {
+    const ctypes = protocol.helpers.parseReturnedContentType(
+        httpReply.headers['content-type']);
+    const size = ctypes.get('data');
+    const expectedCRC = ctypes.get('$crc.data');
+
+    if (!size || !expectedCRC) {
+        return httpReply;
+    }
+
+    let bytesUntilCRC = size;
+    let endDataBuffer = null;
+    let crcBuffer = null;
+    const corruptSaferStream = new stream.Transform({
+        transform(chunk, encoding, callback) {
+            if (chunk.length < bytesUntilCRC) {
+                this.push(chunk);
+                bytesUntilCRC -= chunk.length;
+                callback();
+                return;
+            }
+
+            endDataBuffer = endDataBuffer === null ?
+                chunk.slice(0, bytesUntilCRC) :
+                Buffer.concat([endDataBuffer, chunk.slice(0, bytesUntilCRC)]);
+
+            crcBuffer = crcBuffer === null ?
+                chunk.slice(bytesUntilCRC) :
+                Buffer.concat([crcBuffer, chunk.slice(bytesUntilCRC)]);
+
+            /* Partial CRC */
+            if (crcBuffer.length < 12) {
+                bytesUntilCRC -= chunk.length;
```

Comment:
I'm not sure I get this part. Is that because we may have a partial piece of data, and thus are expecting the CRC to come in (one of) the next chunk(s) ?

Reply:
Yes! I don't know whether chunk will get me to the end of the stream. I might have only read 2 bytes of CRC payload. Granted, this part lacks explaantions, so used to this protocol I take it for granted everyone knows about it